File: pkgs/self/update.go
package self

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"

	"github.com/gvcgo/goutils/pkgs/gutils"
	"github.com/gvcgo/version-manager/pkgs/conf"
)

const (
	updateScriptName string = "vmr-update"
)

var (
	WinScript  string = `powershell -nop -c "iex(New-Object Net.WebClient).DownloadString('https://gvc.1710717.xyz/proxy/https://raw.githubusercontent.com/gvcgo/version-manager/main/scripts/install.ps1')"`
	UnixScript string = `curl --proto '=https' --tlsv1.2 -sSf https://gvc.1710717.xyz/proxy/https://raw.githubusercontent.com/gvcgo/version-manager/main/scripts/install.sh | sh`
)

func setUpdateForWindows() {
	scriptPath := filepath.Join(conf.GetManagerDir(), fmt.Sprintf("%s.bat", updateScriptName))
	os.WriteFile(scriptPath, []byte(WinScript), os.ModePerm)
}

func setUpdateForUnix() {
	scriptPath := filepath.Join(conf.GetManagerDir(), updateScriptName)
	os.WriteFile(scriptPath, []byte(UnixScript), os.ModePerm)
}

func SetUpdateScript() {
	if runtime.GOOS == gutils.Windows {
		setUpdateForWindows()
	} else {
		setUpdateForUnix()
	}
}

File: rust-vmr/crates/vmr-self/src/update.rs
use std::fs;
use vmr_config::conf::DEFAULT_DOMAIN;

const UPDATE_SCRIPT_NAME: &str = "vmr-update";

/// Write the vmr-update helper script (~/.vmr/vmr-update[.bat|.sh]).
pub fn set_update_script() {
    let work_dir = vmr_config::paths::get_vmr_work_dir();

    #[cfg(windows)]
    {
        let win_script = format!(
            "cd %HOMEPATH%\npowershell -c \"irm https://scripts.{}/windows | iex\"",
            DEFAULT_DOMAIN
        );
        let bat_path = work_dir.join(format!("{}.bat", UPDATE_SCRIPT_NAME));
        fs::write(&bat_path, &win_script).ok();

        let mingw_script = format!("#!/bin/sh\ncd ~\npowershell {}", bat_path.display());
        let mingw_path = work_dir.join(format!("{}.sh", UPDATE_SCRIPT_NAME));
        fs::write(&mingw_path, &mingw_script).ok();
    }

    #[cfg(not(windows))]
    {
        let unix_script = format!(
            "#!/bin/sh\ncd ~\ncurl --proto '=https' --tlsv1.2 -sSf https://scripts.{} | sh",
            DEFAULT_DOMAIN
        );
        let script_path = work_dir.join(UPDATE_SCRIPT_NAME);
        fs::write(&script_path, &unix_script).ok();
    }
}
